### PR TITLE
Solve MTU issues by using table 200 post-encryption using output-mark

### PR DESCRIPTION
**Pull request by lbernail**
Tentative to address: https://github.com/cilium/cilium/issues/9112
This is still WIP (in particular we should not use the netlink fork) but I created the PR to discuss it and test it in other environments if needed (I can provide an image with the changes too).
@jrfastab : let me know when you have some time to look into it

How it works:
- decrease MTU in default routing table to account for encryption overhead for traffic from the node (to use with `encrypt-node` option)
- don't decrease MTU in encryption routing table (200)
- set output-mark in the state used for outgoing traffic to force use of table 200 post-encryption

Implementation:
- support output-mark in the netlink library (hence the new vendored version, which I hope will soon not require a fork, upstream PR here: https://github.com/vishvananda/netlink/pull/484
- set MTU in the route before calling `route.Upsert` and remove mtuConfig from route.Upsert parameters (because the MTU now depends on the context)
- add an outputMark parameter to `ipSecReplaceState` (we only want to set the mark on the state used for outgoing traffic)
- add `AdvMSS` to the route struct (not used yet but we'll need for kernels not supporting output-mark)

I've tested it in our setup and it is looking good.

I still have a weird behavior with fragmentation and I'm not sure what is supposed to happen exactly. Here is an example of what I saw:
1 - send an ICMP packet from the host with a payload triggering fragmentation
2 - packets is reassembled after traversing cilium_host
3 - packet is encrypted and sent to cilium_host
4- if after encryption packet is smaller than physical MTU it is successfully sent
5- if after encryption packet is bigger it is dropped

I assume the packet is fragmented in step 5 and dropped by the eBPF code in cilium_host (not sure). The behavior is different in Ubuntu 1904 (kernel 5.0) where fragmented packets are always dropped (not sure why because fragmention is done (checked on cilium_host) and I can see that packets are reassembled (from netstat IP stats) and encrypted (from state counters)). I think we should avoid fragmentation but if the encrypted packet is too large it should still be fragmented and sent to its destination.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9201)

### CODE REVIEW

**Comment by jrfastab** [comment]
test-me-please

**Comment by lbernail** [comment]
I updated the PR with a few additional changes:

1 - split `ipSecReplaceState` into `ipSecReplaceStateIn` and `ipSecReplaceStateOut`
On a 5.0 kernel if the output-mark for outgoing traffic is set on a state used for incoming traffic, packets are dropped with `XfrmInNoPols`
The first idea (original commit) was to use an `outputMark` parameter in `ipSecReplaceState` and set it to 0 when states were added for incoming traffic.
However, states are shared for incoming/outgoing for `0.0.0.0 => subnets` so this did not work (the states was written by both policies)
Splitting the function in two allows for 2 different states and also allows us to match the expected mark in the state

2- comment out the create of policies/states for `dir in` in `enableSubnetIPsec`
I don't see when these states would be used (no encrypted traffic should ever match `0.0.0.0 => subnets` because outer IPs should be node IPs
Doing this also removed the associated `dir in` policies which leads to simply this for incoming traffic:
```
Policy:
src 0.0.0.0/0 dst <nodeip>/32
	dir in priority 0
	mark 0xd00/0xf00
	tmpl src 0.0.0.0 dst nodeip
		proto esp reqid 1 mode tunnel

State:
src 0.0.0.0 dst <nodeip>
	proto esp spi 0x00000001 reqid 1 mode tunnel
	replay-window 0
	mark 0xd00/0xf00 output-mark 0xd00
	aead rfc4106(gcm(aes)) 0xXXXXXXXXXXXXXXXXXXXX 128
	anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
	sel src 0.0.0.0/0 dst 0.0.0.0/0
```

I've tested this in our setup and everything seems to be working fine (no MTU issue anymore) but I'm not familiar with other setups to be sure of the impacts.

Additional note: I've found the issue with fragmented packets. Packet fragmentation uses the control block and overwrites the tunnel_endpoint. In addition, it seems that on recent Kernels (seen on 5.0) the control block is also used during reassembly with the same consequence. The best approach is probably to drop fragmented packets when encryption is enabled (I'm happy to add this to this PR).

Final comment: output-mark is only supported for kernels 4.14+. What is the target in terms of kernel support?

**Comment by jrfastab** [comment]
We also use fib lookup in these code paths which was added in 4.18 so it seems minimal kernel for subnet feature is 4.18. Creating an issue to document version reqs here, https://github.com/cilium/cilium/issues/9230

File: pkg/datapath/linux/node.go [hunk 1]
```diff
@@ -472,8 +472,8 @@ func (n *linuxNodeHandler) enableSubnetIPsec(v4CIDR, v6CIDR []*net.IPNet) {
 		ipsecIPv6Wildcard := &net.IPNet{IP: net.ParseIP(wildcardIPv6), Mask: net.CIDRMask(0, 0)}
 
 		n.replaceNodeIPSecInRoute(cidr)
-		spi, err = ipsec.UpsertIPsecEndpoint(cidr, ipsecIPv6Wildcard, ipsec.IPSecDirIn)
-		upsertIPsecLog(err, "CNI In IPv6", cidr, ipsecIPv6Wildcard, spi)
+		//spi, err = ipsec.UpsertIPsecEndpoint(cidr, ipsecIPv6Wildcard, ipsec.IPSecDirIn)
+		//upsertIPsecLog(err, "CNI In IPv6", cidr, ipsecIPv6Wildcard, spi)
```
Comment (jrfastab) [inline pkg/datapath/linux/node.go]: Lets break this into two commits one to slit up replaceState into an out and in part and then a second one to drop the extra state. That way we can keep track of exactly what changed later. Also commit messages please so we can remember what this was for next month when reading git log.

File: Gopkg.toml [hunk 2]
```diff
@@ -435,7 +435,8 @@ required = [
 
 [[constraint]]
   name = "github.com/vishvananda/netlink"
-  revision = "b1e98597921720cee4147d6476110fc8fc56072d"
+  source = "github.com/DataDog/netlink.git"
+  branch = "lbernail/output-mark"
```
Comment (jrfastab) [inline Gopkg.toml]: next patch  pushes this to correct revision, so just put that change here vs changing and then changing it to correct value later.

Comment (jrfastab) [reply Gopkg.toml]: my preference would be to update the netlink library in its own commit.

File: Gopkg.toml [hunk 3]
```diff
@@ -435,8 +435,7 @@ required = [
 
 [[constraint]]
   name = "github.com/vishvananda/netlink"
-  source = "github.com/DataDog/netlink.git"
-  branch = "lbernail/output-mark"
+  revision = "e906d226249b76de7fe3c326ea4616ca3b5945a6"
```
Comment (jrfastab) [inline Gopkg.toml:438]: move this patch to front

File: pkg/datapath/linux/node.go (hunk 1 above)
Comment (jrfastab) [reply pkg/datapath/linux/node.go]: also just delete the lines assume that is just a wip thing though.

**Comment by jrfastab** [comment]
Overall looks good to me, clean it up and it should be good to go in.

**Review by jrfastab — APPROVED** [review]
LGTM

**Comment by lbernail** [comment]
It seems the PR broke something:
```
time="2019-09-19T16:21:47Z" level=debug msg="running command: kubectl exec -n default testds-4m74c -- ping -W 2 -c 5 10.10.0.24"
time="2019-09-19T16:21:54Z" level=error msg="Error executing command 'kubectl exec -n default testds-4m74c -- ping -W 2 -c 5 10.10.0.24'" error="exit status 1"
time="2019-09-19T16:21:54Z" level=error msg="Error executing command 'kubectl exec -n default testds-4m74c -- ping -W 2 -c 5 10.10.0.24'" error="exit status 1"
cmd: "kubectl exec -n default testds-4m74c -- ping -W 2 -c 5 10.10.0.24" exitCode: 1 duration: 7.214452081s stdout:
PING 10.10.0.24 (10.10.0.24): 56 data bytes
--- 10.10.0.24 ping statistics ---
5 packets transmitted, 0 packets received, 100% packet loss

stderr:
command terminated with exit code 1```

The configuration that fails seems to be direct_routing with encryption (close to what we use except we also use subnets). Is there a way to replicate the test environment so I can see what is wrong?

**Comment by lbernail** [comment]
Latest push: fix a test using the previous route.Upsert() method signature (sorry I missed it before)

File: vagrant_box_defaults.rb [hunk 4]
```diff
@@ -2,6 +2,8 @@
 # vi: set ft=ruby
 Vagrant.require_version ">= 2.2.0"
 
+#$SERVER_BOX = "cilium/ubuntu-next"
```
[the remark this answers is not in the document]
Comment (lbernail) [reply vagrant_box_defaults.rb]: Sorry I pushed some local hacks by mistake in the first commit...

**Comment by jrfastab** [comment]
test-me-please

**Comment by lbernail** [comment]
I fixed the unit test (I forgot to remove the import...)
I can rearrange the commits when the tests are ok

**Comment by lbernail** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

File: Gopkg.toml [hunk 5]
```diff
@@ -435,7 +435,7 @@ required = [
 
 [[constraint]]
```
Comment (ianvernon) [inline Gopkg.toml:436]: Please sign off on your commit.

**Review by ianvernon — CHANGES_REQUESTED** [review]
Please sign off on all of your commits. I will defer approval to @jrfastab regarding the actual code changes.

**Comment by jrfastab** [comment]
code looks good to me, one small nit though.

Can we squash `Use output-mark to use table 200 post-encryption and set different MT… ` `Only set output mark for encrypt-subnet and encrypt-node` into one commit so we can bisect this if we ever need to. And like Ian said add signed-off-by.

If it gets too hard to squash them for whatever reason let me know.

**Comment by lbernail** [comment]
@jrfastab : I'll do my best to squash the commits (and also squash the last fix of course)
Sorry I forgot the signoff option

**Comment by lbernail** [comment]
I managed to squash the commits without any issue and added the signoff to commit messages

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
retest seems we hit this https://github.com/cilium/cilium/issues/8775? 

Logs here, https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/15171/testReport/junit/Suite-k8s-1/11/K8sIstioTest_Istio_Bookinfo_Demo_Tests_bookinfo_inter_service_connectivity/

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
LGTM and @ianvernon request is resolved.

**Comment by tgraf** [comment]
Nice work everyone!

**Pull request merged**